### ISSUE TITLE
Question about cuda kernel performance

### ISSUE THREAD

**Issue by kehuanfeng** [tensorflow/tensorflow#58796]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

source

### Tensorflow Version

tf 2.11.0

### Custom Code

Yes

### OS Platform and Distribution

Linux/Ubuntu 20.04

### Mobile device

no

### Python version

3.8.10

### CUDA/cuDNN version

cuda V11.2.152/cudnn 8.1.0

### Current Behaviour?

When running my model in eager mode on NVIDIA T4, The elapsed time for cuda kernel will become faster after several iterations (4 to be precise).

The following is the summary from Nsight system. 
As shown, for eager mode, the max is 1.2ms, while the avg is about 500 us after 100 iterations;
for compat.v1.Session style with the same code base, the kernel time is about 1.2ms consistently.  

The question is how does it happen, and which feature from tf2/eager mode can make cuda kernel run faster after several iters? As I understand, the cuda kernel is either from cudnn or cublas, tf itself or tracing or graph compilation should not touch that low-level logic.

```shell
// TF2.11.0 eager mode
 Time (%)  Total Time (ns)  Instances  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)     GridXYZ         BlockXYZ                                                     Name
 --------  ---------------  ---------  --------  --------  --------  --------  -----------  --------------  --------------  ----------------------------------------------------------------------------------------------------
     13.4        258681374        500  517362.7  479449.5    469753   1222701     138505.1     8    5    1   256    1    1  volta_sgemm_128x128_nn 


// TF 2.11.0 tf.compat.v1.Session()
 Time (%)  Total Time (ns)  Instances   Avg (ns)     Med (ns)    Min (ns)   Max (ns)   StdDev (ns)     GridXYZ         BlockXYZ                                                     Name
 --------  ---------------  ---------  -----------  -----------  ---------  ---------  -----------  --------------  --------------  ----------------------------------------------------------------------------------------------------
     16.4      607,754,406        500  1,215,508.8  1,221,934.0  1,161,743  1,228,110     18,653.6     8    5    1   256    1    1  volta_sgemm_128x128_nn
```

### Standalone code to reproduce the issue

```shell
// eager mode
for i in tf.range(100):
    tf.print(model(x))

// sess
With tf.compat.v1.Session() as sess:
     print(sess.run(output_name_list, x))
```

### Relevant log output

_No response_</details>

**Comment by kehuanfeng** [comment]
Hi @mohantym,

Thanks for the reply! I couldn't share my private model so I made a simple example as below,
```
import argparse
import tensorflow as tf
import timeit

parser = argparse.ArgumentParser(description='TensorFlow bench',
                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument('--eager', action='store_true', default=False,
                    help='Enable eager execution')

args = parser.parse_args()
if not args.eager:
    tf.compat.v1.disable_eager_execution()

data = tf.zeros([640, 4096])
model = tf.keras.Sequential([
    tf.keras.layers.Dense(2048),
    tf.keras.layers.Dense(256),
    tf.keras.layers.Dense(10),
])

if args.eager:
    print(timeit.timeit(lambda: model(data), number=100))
else:
    with tf.compat.v1.Session() as sess:
        sess.run(tf.compat.v1.global_variables_initializer())
        print(timeit.timeit(lambda: model.predict(data, steps=1), number=100))
```

You can run it, 
```
nsys profile python main.py --eager
```
and check the cuda kernel time distribution.

(Not sure why tf1 sess picks up the different kernels compared to tf2. Here we can just focus on tf2 kernel time)

**Comment by mohantym** [comment]
Ok @kehuanfeng !
Thanks for the update.

@SuryanarayanaY !
Could you look at this issue.

Thank you!

**Comment by SuryanarayanaY** [comment]
Hi @kehuanfeng ,

Actually for larger models Graph mode can give better performance than eager mode.From TF 2.x onwards Eager mode made as default due to beginner friendly, easy to Debug etc and Graph mode as optional by using `tf.function`. I am not comparing 1.X Graphs here since we are not supporting 1.X versions now. Iam comparing performance of a function with larger epochs running on Eager mode and Graph mode where we can see Graph mode is clearly a winner. Please refer to the attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/ea3490ee50b4dbaec4eec641d0962885/58796-r1.ipynb).

Thankyou!

**Comment by kehuanfeng** [comment]
Hi @SuryanarayanaY, @mohantym,

Thanks for the reply.
I updated the test sample as below,
```
import argparse
import tensorflow as tf
import timeit

parser = argparse.ArgumentParser(description='TensorFlow bench',
                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument('--mode',
                    choices=['eager', 'tf2graph', 'tf1graph'],
                    help='execution mode')

args = parser.parse_args()

if args.mode == "tf1graph":
     tf.compat.v1.disable_eager_execution()

data = tf.zeros([640, 4096])
model = tf.keras.Sequential([
    tf.keras.layers.Dense(2048),
    tf.keras.layers.Dense(256),
    tf.keras.layers.Dense(10),
])

if args.mode == "eager":
    def run_model():
        return model(data)
    run_model()
    print(timeit.timeit(lambda: run_model(), number=1000))
elif args.mode == "tf2graph":
    @tf.function()
    def run_model():
        return model(data)
    run_model()
    print(timeit.timeit(lambda: run_model(), number=1000))
elif args.mode == "tf1graph":
    with tf.compat.v1.Session() as sess:
        sess.run(tf.compat.v1.global_variables_initializer())
        model.predict(data, steps=1)
        print(timeit.timeit(lambda: model.predict(data, steps=1), number=1000))
(venv) 
```
I am trying to compare the performance of eager mode, tf2 graph mode and tf1 graph mode. And to be more precise,  I add warm-up logic for each mode.

I got the following result for 1000 iterations for each mode,
```
TF1 graph：1.027 s
TF2 graph：1.871 s
TF2 eager：1.926 s (not very stable)
```
As you can see, TF1 graph mode is the fastest one and the eager mode is the slowest one. 

And then I profiled the cuda kernel,

```
 TF1 graph：
 
 Time (%)  Total Time (ns)  Instances  Avg (ns)   Med (ns)   Min (ns)  Max (ns)  StdDev (ns)     GridXYZ         BlockXYZ                                                     Name                                                
 --------  ---------------  ---------  ---------  ---------  --------  --------  -----------  --------------  --------------  ----------------------------------------------------------------------------------------------------
     80.2      258,716,663      1,001  258,458.2  258,461.0   257,950   259,325        181.9     2   10    4   128    1    1  volta_sgemm_128x64_nn                                                                               
      8.5       27,481,458      1,001   27,454.0   27,424.0    27,232    28,064        110.4    40    1    1  1024    1    1  void Eigen::internal::EigenMetaKernel<Eigen::TensorEvaluator<const Eigen::TensorAssignOp<Eigen::Ten…
      6.7       21,734,422      1,001   21,712.7   21,696.0    21,568    22,271         82.0     1   20    1   128    1    1  volta_sgemm_32x32_sliced1x4_nn                                                                      
      2.8        8,932,829      1,001    8,923.9    8,928.0     8,768     9,344         61.0    40    1    1  1024    1    1  void tensorflow::BiasNHWCKernel<float>(int, const T1 *, const T1 *, T1 *, int)                      
      1.5        4,921,638      1,001    4,916.7    4,896.0     4,863     5,280         48.5     7    1    1  1024    1    1  void tensorflow::BiasNHWCKernel<float>(int, const T1 *, const T1 *, T1 *, int) 
	  
	  
TF2 graph：	  

 Time (%)  Total Time (ns)  Instances   Avg (ns)     Med (ns)    Min (ns)   Max (ns)   StdDev (ns)     GridXYZ         BlockXYZ                                                     Name                                                
 --------  ---------------  ---------  -----------  -----------  ---------  ---------  -----------  --------------  --------------  ----------------------------------------------------------------------------------------------------
     90.8    1,946,823,575      1,001  1,944,878.7  1,867,018.0  1,789,100  3,779,700    371,002.2    16   10    8   128    1    1  volta_sgemm_128x64_nn                                                                               
      6.4      136,219,164      1,001    136,083.1    130,847.0    125,375    259,165     24,835.7     2   10    4   128    1    1  volta_sgemm_128x64_nn                                                                               
      2.1       45,219,493      2,002     22,587.2     23,727.5      5,376     48,224     16,882.3    40    1    1  1024    1    1  void tensorflow::BiasNHWCKernel<float>(int, const T1 *, const T1 *, T1 *, int)


TF2 eager：

 Time (%)  Total Time (ns)  Instances   Avg (ns)     Med (ns)    Min (ns)   Max (ns)   StdDev (ns)     GridXYZ         BlockXYZ                                                     Name                                                
 --------  ---------------  ---------  -----------  -----------  ---------  ---------  -----------  --------------  --------------  ----------------------------------------------------------------------------------------------------
     90.8    1,976,980,767      1,001  1,975,005.8  1,848,266.0  1,757,708  3,780,050    457,204.1    16   10    8   128    1    1  volta_sgemm_128x64_nn                                                                               
      6.3      137,923,197      1,001    137,785.4    129,407.0    123,358    259,293     30,686.3     2   10    4   128    1    1  volta_sgemm_128x64_nn                                                                               
      2.1       45,871,552      2,002     22,912.9     23,984.0      5,280     50,400     17,175.4    40    1    1  1024    1    1  void tensorflow::BiasNHWCKernel<float>(int, const T1 *, const T1 *, T1 *, int)
	  
```
The sgemm kernel takes the most time.  
If we step into the 'Avg', 'Min', 'Max' time of sgemm kernel for each mode, we can see tf2 kernel time is not stable, regardless eager or graph mode. Is there any reason for this behavior?

Actually, if you check the profiling result through nsight gui, you will find the kernel runs slowly in the begining and  become faster after several iterations.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
